### PR TITLE
Filter out stake and vote accounts with incorrect owners

### PR DESCRIPTION
**Pull request by CriesofCarrots**
#### Problem
Follow up to #13615
While the stake program now checks whether a vote account is correctly owned on Delegation, the runtime method to collect stake and vote accounts doesn't verify whether this is still the case.

#### Summary of Changes
Filter out invalid stake and vote accounts

### CODE REVIEW

File: runtime/src/bank.rs [hunk 1]
```diff
@@ -1627,19 +1628,27 @@ impl Bank {
...
+                                    vote_account.owner,
                                 ),
                             ));
                         }
+                        if self
+                            .feature_set
+                            .is_active(&feature_set::filter_stake_delegation_accounts::id())
+                            && (stake_account.owner != solana_stake_program::id()
+                                || vote_account.owner != solana_vote_program::id())
+                        {
```
Comment (CriesofCarrots) [inline runtime/src/bank.rs:1646]: @ryoqun what reporting do you think would be most helpful here?

File: runtime/src/bank.rs [hunk 2]
```diff
@@ -1610,6 +1610,7 @@ impl Bank {
     ///  returns a map (has to be copied) of loaded
     ///   ( Vec<(staker info)> (voter account) ) keyed by voter pubkey
     ///
+    /// Filters out invalid pairs
```
Comment (ryoqun) [inline runtime/src/bank.rs:1613]: :+1:

File: runtime/src/bank.rs [hunk 3]
```diff
@@ -11305,4 +11314,53 @@ pub(crate) mod tests {
...
+            .unwrap_or_default();
+        stake_account.owner = bogus_stake_program;
+        bank.store_account(
+            &validator_vote_keypairs1.stake_keypair.pubkey(),
+            &stake_account,
+        );
+
+        // Accounts must be valid stake and vote accounts
+        let stake_delegation_accounts = bank.stake_delegation_accounts(&mut null_tracer());
+        assert_eq!(stake_delegation_accounts.len(), 0);
```
Comment (ryoqun) [inline runtime/src/bank.rs:11364]: :+1:

File: runtime/src/bank.rs (hunk 1 above)
Comment (ryoqun) [reply runtime/src/bank.rs:1646]: hmm, I think just `error!()` suffices. At least this is more significant than `warn!`, imo.

Comment (CriesofCarrots) [reply runtime/src/bank.rs:1646]: That seems a bit strong to me; it's not really a node or runtime error, is it? Would a metric datapoint reporting the slot and the address perhaps be more useful?

File: runtime/src/bank.rs [hunk 4]
```diff
@@ -11305,4 +11314,53 @@ pub(crate) mod tests {
...
+        assert_eq!(stake_delegation_accounts.len(), 2);
+
+        // Modify staked vote account owner; a vote account owned by another program could be
+        // freely modified with malicious data
+        let bogus_vote_program = Pubkey::new_unique();
+        let mut vote_account = bank
+            .get_account(&validator_vote_keypairs0.vote_keypair.pubkey())
+            .unwrap_or_default();
+        vote_account.owner = bogus_vote_program;
+        bank.store_account(
```
Comment (ryoqun) [inline runtime/src/bank.rs:11343]: I think runtime enforces an intermediately write of account.lamports == 0 when chainging existing non-system accounts.
So, how about inserting this here?:

```rust
// simulate vote account removal via full withdrawal
bank.store_account(&validator_vote_keypairs0.vote_keypair.pubkey(), &Acount::default());
```

Then, the following assertion should be `1`: 

```rust
assert_eq!(bank.vote_accounts().len(), 1);
```

File: runtime/src/bank.rs (hunk 1 above)
Comment (ryoqun) [reply runtime/src/bank.rs:1646]: > That seems a bit strong to me; it's not really a node or runtime error, is it? Would a metric datapoint reporting the slot and the address perhaps be more useful?

Hhhm, yeah, I agree. After more thought, I think we can go with your suggested approach.

**Comment by ryoqun** [comment]
@CriesofCarrots btw, did you tested this manually?:

Well, I have a secret sauce of easily creating fake account via the wracked `solana deploy` command:

``` bash
$ ./target/release/solana --url http://localhost:8899 account -o vote.data <foo bar pubkey json>
$ hexedit
$ ./target/release/solana.bad-deploy deploy --url http://localhost:8899/ vote.data <foo bar pubkey json>
```

```
diff --git a/cli/src/cli.rs b/cli/src/cli.rs
index aa7b69fb7c..e1ee8aa66b 100644
--- a/cli/src/cli.rs
+++ b/cli/src/cli.rs
@@ -1227,8 +1227,8 @@ fn do_process_deploy(
         CliError::DynamicProgramError(format!("Unable to read program file: {}", err))
     })?;
 
-    Executable::<BPFError>::from_elf(&program_data, Some(|x| bpf_verifier::check(x, true)))
-        .map_err(|err| CliError::DynamicProgramError(format!("ELF error: {}", err)))?;
+    //Executable::<BPFError>::from_elf(&program_data, Some(|x| bpf_verifier::check(x, true)))
+    //    .map_err(|err| CliError::DynamicProgramError(format!("ELF error: {}", err)))?;
 
     let loader_id = if use_deprecated_loader {
         bpf_loader_deprecated::id()

```

this is not strict requirement, though. :)

**Review by ryoqun — DISMISSED** [review]
lgtm after metrics addition and small test tweaks!

Really thanks for taking this over. :)

**Comment by ryoqun** [comment]
> Well, I have a secret sauce...

@CriesofCarrots And this for faster inflation:


```
SOLANA_RUN_SH_CLUSTER_TYPE=mainnet-beta/development SOLANA_RUN_SH_GENESIS_ARGS="--slots-per-epoch 128 --vote-commission-percentage=55" ./run.sh
```

File: runtime/src/bank.rs (hunk 4 above)
Comment (CriesofCarrots) [reply runtime/src/bank.rs:11343]: That is more realistic, but it doesn't change `bank.vote_accounts().len()`. (Confirmed with test run.)
Unless I'm missing something, I don't see that Stakes ever removes vote_accounts, regardless of owner or lamports balance.

Comment (ryoqun) [reply runtime/src/bank.rs:11353]: well, sorry, my comment was partly wrong.

Account::default().owner isn't Vote1111; that's why it didn't change.

So, the correct code should be like this:

```
vote_account.lamports = 0;
// simulate vote account removal via full withdrawal
bank.store_account(&validator_vote_keypairs0.vote_keypair.pubkey(), &vote_account);

vote_account.lamports = original_lamports;
vote_account.owner = bogus_vote_program;

bank.store_account(&validator_vote_keypairs0.vote_keypair.pubkey(), &vote_account);
assert_eq!(bank.vote_accounts().len(), 1);

```

Well, Stake::store is hard to read without much aid; Please check this out: https://github.com/solana-labs/solana/pull/14065/files#r540752554 I still think it should remove `vote_accounts`.

Comment (CriesofCarrots) [reply runtime/src/bank.rs:11353]: Oh thank you, I did miss that all owned vote accounts are removed and then potentially re-added.
Okay, great, your suggested code looks just like mine does now :)

**Comment by CriesofCarrots** [comment]
@ryoqun thanks for the special sauce! It took a few more steps, but I was able to test this manually. Merging!

**Pull request merged**